### PR TITLE
(devex)(error-tracking) Generate JS source-maps for the notifications app

### PR DESCRIPTION
**Pull request by fullofcaffeine**
## Proposed Changes

Generate source-maps for the `notifications` _standalone_ app JS assets. When `NODE_ENV` is _not_ `production`, it will use `inline-cheap-source-map` as `devtool`. For production builds, `source-map` will be used, following the same pattern of other Calypso apps (such as ETK).

### Production build

1. Run `yarn build`
2. Verify `dist/build.min.js.map` is present (see screenshot below):
Screenshot from 2023-06-12 17-57-22

#### Verify that source-maps are applied in production by sandboxing `widgets.wp.com`:

1. Sandbox `widgets.wp.com`;
2. Run `yarn build`
3. `cd dist` and then run `rsync -azP  --exclude .git --exclude --delete user@sandbox:~/public_html/widgets.wp.com/notifications`
4. Open the notifications panel by clicking the bell icon on the upper-right corner of the screen
5. Verify source-maps have been applied:

Screenshot from 2023-06-12 18-27-10

This tests that source-maps have been applied to the standalone app.

#### Verify source-maps are applied in production in the `wp-calypso` web app:

Source-maps are already being generated for the Calypso web app. The notification is rendered within the Calypso client, which already has source-map generation setup (see https://github.com/Automattic/wp-calypso/pull/62965). It uses `hidden-source-maps`, so they are not loaded by the devtools and we can't easily verify, but we'll trust it's already working :) '

### Development mode

1. Run `yarn dev`
7. Open `dist/build.min.js`, scroll to the bottom and notice the inline source-map:
Screenshot from 2023-06-12 17-56-27

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: apps/notifications/webpack.config.js [hunk 1]
```diff
@@ -102,6 +103,7 @@ function getWebpackConfig(
 			new ExtensiveLodashReplacementPlugin(),
 			BuildMetaPlugin( { outputPath } ),
 		].filter( Boolean ),
+		devtool: isDevelopment ? 'inline-cheap-source-map' : 'source-map',
```
Comment (fullofcaffeine) [inline apps/notifications/webpack.config.js:112]: A couple of Calypso apps use `source-map` as the `devtool` on production, but the Calypso client uses `hidden-source-map`. I think from the perspective of devex, `source-map` is better, while from a perf. perspective, one could argue `hidden-source-maps` saves some bytes. I decided to leave it as `source-map`, but happy to hear other people's thoughts about it, if any!

**Comment by noahtallen** [comment]
> cd dist and then run rsync -azP  --exclude .git --exclude --delete user@sandbox:~/public_html/widgets.wp.com/notifications

By the way, there are two built-in ways to do this:

1. `yarn build --sync`
2. `install-plugin.sh notifications generate/js-source-maps-for-notifications-app` (this would more closely match the production release asset!)

I used the install-plugin approach, but I'm noticing the map file isn't getting applied:

```sh
~/public_html (trunk) $ install-plugin.sh notifications generate/js-source-maps-for-notifications-app
Downloading...
Unzipping...
Checking for unsafe deploy...
Updating files with version control...

Build info:
  - TeamCity build number: 1.65272
  - Commit hash: 171fb9e
  - Commit link: https://github.com/Automattic/wp-calypso/commit/171fb9e0181e64542fb0cd7811458259e1b3f60c
  - Direct download: https://teamcity.a8c.com/repository/download/calypso_WPComPlugins_Notifications/notifications-release-build.tcbuildtag/notifications.zip?guest=1&branch=generate/js-source-maps-for-notifications-app

Successfully installed Notifications build 1.65272
~/public_html (trunk) $ git status
On branch trunk
Your branch is up to date with 'origin/trunk'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   widgets.wp.com/notifications/build.min.css
	modified:   widgets.wp.com/notifications/build.min.js
	modified:   widgets.wp.com/notifications/build.min.js.LICENSE.txt
	modified:   widgets.wp.com/notifications/build.min.rtl.css
	modified:   widgets.wp.com/notifications/build_meta.json
	modified:   widgets.wp.com/notifications/index.html
	modified:   widgets.wp.com/notifications/rtl.html
	modified:   wp-content/lib/calypso-apps/cache-busters.json
```

Even though the `.min.js.map` file is in the buildfile. It's possible we also need to update the install-script to allow .map files?

**Comment by fullofcaffeine** [comment]
> Even though the .min.js.map file is in the buildfile. It's possible we also need to update the install-script to allow .map files?

~~Interesting! Thanks for letting me know, I'll look into that!~~ [See comment below](https://github.com/Automattic/wp-calypso/pull/78081#issuecomment-1588363218).

**Comment by fullofcaffeine** [comment]
> yarn build --sync

Ahh, nice. I knew about `dev --sync` but forgot about `build --sync`, thanks for the nudge :)

**Comment by fullofcaffeine** [comment]
> Interesting! Thanks for letting me know, I'll look into that!

Ah, I think I know what's happening. We do remove .map files from production and upload them to Sentry using [this build](https://teamcity.a8c.com/buildConfiguration/calypso_WPComPlugins_GutenbergUploadSourceMapsToSentry/10191998?buildTab=log), so that's expected. That means it might make sense to change the `devtool` to `hidden-source-map` OR to adjust the `install-plugin.sh` script to allow pushing source-maps. However - if my memory doesn't fail - we remove source-maps in the `install-plugin.sh` to reduce the size of the commit and reduce the risk of causing a `hunt suc dracones` situation (the svn commit size limit error message).

**Comment by noahtallen** [comment]
Ah, forgot about that!

> However - if my memory doesn't fail - we remove source-maps in the install-plugin.sh to reduce the size of the commit

I don't think the size issue should be a problem for notifications (ETK could be a different story). So if we need to push it to work around the issue, that'd be fine. However, if we want to match the ETK+Gutenberg behavior and add notifications source maps to that build too, hidden source map sounds good

**Comment by fullofcaffeine** [comment]
For uploading soruce-maps to Sentry for the `notifications` app, it _should_ work as is. The TC build for `notifications` should contain a complete build with all assets, including source-maps, no? 

I'd love to commit source-maps to the SVN repo (so that users and devs could see source-mapped JS assets in the devtools), but given the SVN size limit, I'll probably refrain from doing so atm.

**Comment by fullofcaffeine** [comment]
> For uploading soruce-maps to Sentry for notifications, it should work as is. The TC build for notifications should contain a complete build with all assets, including source-maps, no?

See:
* https://github.com/Automattic/wp-calypso/blob/trunk/.teamcity/_self/lib/wpcom/WPComPluginBuild.kt#L114
* https://github.com/Automattic/wp-calypso/blob/trunk/apps/notifications/package.json#L24
* https://github.com/Automattic/wp-calypso/blob/trunk/.teamcity/_self/projects/WPComPlugins.kt#LL409C2-L409C2

**Comment by noahtallen** [comment]
Yep! The artifact does have the map file:

image

**Review by noahtallen — APPROVED** [review]
What you mentioned sounds good to me 👍

File: apps/notifications/webpack.config.js (hunk 1 above)
Comment (noahtallen) [reply apps/notifications/webpack.config.js:112]: Another point is that you get warnings in the browser console when a file includes a link to a source map, but one isn't served from the same location. If it doesn't make a difference to Sentry, we could just use hidden source map for all the Calypso apps

**Comment by fullofcaffeine** [comment]
There's a related discussion about capturing exceptions generated by the standalone `notifications` app here: p1686680165242869-slack-C7YPUHBB2. @noahtallen brought up a good point about errors generated from `notifications` not having access to any error-tracking handling because it runs from an iframe. This shouldn't block the merge of this PR, but we'll need at least two other PRs for error-tracking to start working on the `notifications` standalone app:
1. A PR that exposes error-tracking to the WPCOM simple frontend
1. A PR that adds logic in the notifications standalone app to capture errors and forward them through the iframe proxy to the parent frame, so that the Sentry instance in the WPAdmin context (exposed by the ETK) can pick it up. It will be recorded in the `gutenberg-wp-admin` Sentry project. cc @noahtallen 
1. The backend PHP for the error-tracking should only activate Sentry for logged-in users. There's no such guard right now, but since we use the `admin_` actions, Sentry is never really loaded in the frontend.

**Pull request merged**